### PR TITLE
Backend: Refactor fare policy and fare parameters

### PR DESCRIPTION
**Pull request by DonDokidon**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates

### Additional Changes

- [x] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables
- [x] This PR contains API breaking changes

## How did you test it?

manually

## Checklist

- [x] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

**Review by hkmangla — CHANGES_REQUESTED** [review]

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Tabular/FarePolicy/ProgressiveFarePolicyDetails.hs [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import Kernel.Types.Common (HighPrecMoney, Meters)
+import Kernel.Types.Id
+import Storage.Tabular.FarePolicy (FarePolicyTId)
+
+mkPersist
+  defaultSqlSettings
+  [defaultQQ|
+    ProgressiveFarePolicyDetailsT sql=fare_policy_progressive_details
+      farePolicyId FarePolicyTId
+      startMeters Meters
```
[the remark this answers is not in the document]
Comment (DonDokidon) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Tabular/FarePolicy/ProgressiveFarePolicyDetails.hs:35]: > Are we using this field anywhere?

https://github.com/nammayatri/nammayatri/pull/718/files#diff-c1734e1a28b22d9585cad29615a4e3121da2a4881ea3d3f70b42c96b19406ddcR85

[the remark this answers is not in the document]
Comment (DonDokidon) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Tabular/FarePolicy/ProgressiveFarePolicyDetails.hs:35]: > Are we using this field anywhere?

Yes, that's supposed to be `baseDistance`

File: Backend/dev/migrations/dynamic-offer-driver-app/0157-refactor-fare-policy.sql [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+--      T1.vehicle_variant,
+--      'Slabs',
+--      T1.night_shift_start,
+--      T1.night_shift_end,
+--      T1.max_allowed_trip_distance,
+--      T1.min_allowed_trip_distance,
+--      T1.service_charge,
+--      T1.govt_charges)
+--     FROM atlas_driver_offer_bpp.slab_fare_policy AS T1 );
+-- Slabs are saved using show/read. It's rather difficult to migrate them
```
Comment (DonDokidon) [inline Backend/dev/migrations/dynamic-offer-driver-app/0157-refactor-fare-policy.sql:140]: Slab fare policy are not migrated. You have to migrate them manually or just create new fare policies for them

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Booking.hs [hunk 3]
```diff
@@ -24,68 +25,41 @@ import Kernel.Storage.Esqueleto as Esq hiding (findById, isNothing)
...
-    `innerJoin` table @Fare.FareParametersT
-      `Esq.on` ( \(rb :& _ :& _ :& farePars) ->
-                   rb ^. BookingFareParametersId ==. farePars ^. Fare.FareParametersTId
-               )
-
 findById :: Transactionable m => Id Booking -> m (Maybe Booking)
-findById bookingId = buildDType $
-  fmap (fmap $ extractSolidType @Booking) $
+findById bookingId = buildDType $ do
+  res <-
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Booking.hs:42]: why are these changes?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/DriverQuote.hs [hunk 4]
```diff
@@ -45,13 +50,13 @@ baseDriverQuoteQuery =
                )
 
 findById :: (Transactionable m) => Id Domain.DriverQuote -> m (Maybe Domain.DriverQuote)
-findById dQuoteId = buildDType $
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/DriverQuote.hs:48]: why are these changes?

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/FarePolicy/FarePolicySlabsDetails/FarePolicySlabsDetailsSlab.hs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+
+findAll' ::
+  ( Transactionable m,
+    Monad m,
+    MonadThrow m,
+    Log m
+  ) =>
+  Id DFP.FarePolicy ->
+  DTypeBuilder m [FarePolicySlabsDetailsSlabT]
+findAll' farePolicyId = do
```
Comment (hkmangla) [inline Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/FarePolicy/FarePolicySlabsDetails/FarePolicySlabsDetailsSlab.hs:32]: We should have CachedQueries too for this

**Review by hkmangla — CHANGES_REQUESTED** [review]

**Comment by hkmangla** [comment]
Hey Please add backend or frontend tag before commit message and PR tille.

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/FarePolicy/FarePolicySlabsDetails/FarePolicySlabsDetailsSlab.hs (hunk 5 above)
Comment (DonDokidon) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/FarePolicy/FarePolicySlabsDetails/FarePolicySlabsDetailsSlab.hs:32]: > We should have CachedQueries too for this

For what? We never use it as standalone fetch, it is always a part of FarePolicy, which is already cached

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/DriverQuote.hs (hunk 4 above)
Comment (DonDokidon) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/DriverQuote.hs:48]: > why are these changes?

To be able to build fare parameters

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Booking.hs (hunk 3 above)
Comment (DonDokidon) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/Storage/Queries/Booking.hs:42]: > why are these changes?

same

File: Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FareCalculator.hs [hunk 6]
```diff
@@ -12,6 +12,195 @@
...
+import Domain.Types.FareParameters
+import qualified Domain.Types.FareParameters as DFParams
+import Domain.Types.FarePolicy
+import qualified Domain.Types.FarePolicy as DFP
+import EulerHS.Prelude hiding (id)
+import Kernel.Prelude
+import Kernel.Utils.Common
+
+mkBreakupList :: (Money -> breakupItemPrice) -> (Text -> breakupItemPrice -> breakupItem) -> FareParameters -> [breakupItem]
+mkBreakupList mkPrice mkBreakupItem fareParams = do
```
[the remark this answers is not in the document]
Comment (DonDokidon) [reply Backend/app/provider-platform/dynamic-offer-driver-app/Main/src/SharedLogic/FareCalculator.hs:41]: >  As at estimate stage, we send driver_min_fee, driver_max_fee also to show to customer that what all amount driver can add. 

Does it have to be in breakups? It doesn't seem like breakup to me, can i pass it via something else? Like tags or anything? Also do i have to send it at ride end?

> But at final fare stage, we send driver added fare.

It won't be included if it is Nothing, so this is not a problem at all

**Review by hkmangla — APPROVED** [review]

**Pull request merged**